DEP: io.mmwrite: deprecate `precision` > 15

AtsushiSakai opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Fix #19949.

#### What does this implement/fix?
<!--Please explain your changes.-->
I added the `precision` argument validation for `mmwrite` because the argument value must be in [1, 15] when using float 64.
I added this validation for both C++ implementation and legacy python one.
Also, docstring was updated, and added some unit tests.

sagi-ezri committed:
add fwind1 filr filter on 2d

jakevdp committed:
ENH: scipy.signal.window: add array API support

jakevdp committed:
TST: array API support in test_windows

sagi-ezri committed:
fix circular

sagi-ezri committed:
fix lint

sagi-ezri committed:
update all

sagi-ezri committed:
add to __all__

sagi-ezri committed:
update docstring

sagi-ezri committed:
fix

sagi-ezri committed:
edit docstring

sagi-ezri committed:
update filter

sagi-ezri committed:
fix

sagi-ezri committed:
fix

sagi-ezri committed:
update

sagi-ezri committed:
fix lint

sagi-ezri committed:
edit

sagi-ezri committed:
edit

sagi-ezri committed:
update

sagi-ezri committed:
fix bug

j-towns committed:
MAINT: stats.dirichlet_multinomial relax n to >= 0

Also update docstring and tests accordingly. Some discussion and
motivation for this change: https://github.com/scipy/scipy/issues/21531.

sagi-ezri committed:
fix comments

sagi-ezri committed:
bump

sagi-ezri committed:
update tests

DietBru committed:
Fixed docstring examples.

DietBru committed:
Fixed numeric output docstring examples.

sagi-ezri committed:
update

sagi-ezri committed:
Merge branch 'add-fwind1-filter' of github.com:sagi-ezri/scipy into add-fwind1-filter

sagi-ezri committed:
fix

sagi-ezri committed:
resolve fc

sagi-ezri committed:
fix

sagi-ezri committed:
refactor function name and inputs

sagi-ezri committed:
fix docstring

sagi-ezri committed:
undo examples removal

sagi-ezri committed:
fix lint

DietBru committed:
Fix docstring and unit test.

ev-br committed:
MAINT: signal/windows: assorted fixes to address the review

ev-br committed:
MAINT: signal/widows: continue adding support for Array API

ev-br committed:
TST: signal: convert test_splines.py

symiirorder{1,2} are numpy-only compiled code, thus
- the public objects convert to numpy on entry, convert back on exit
- tests are heavily decorated with `np_only=True` (for private C-impl objects), or
  `cpu_only=True, exceptions=["cupy"]`

ev-br committed:
TST: signal/windows: add explicit dtypes to asarrays

ev-br committed:
TST: signal: convert test_bsplines.py

ev-br committed:
MAINT: signal: make symiirorder{1,2} tests pass on CuPY

ev-br committed:
TST: signal: make sure tests pass with SCIPY_DEVICE=cuda

ev-br committed:
appease the linter

ev-br committed:
TST: signal/windows: fix tests

AtsushiSakai committed:
BUG: io: add `precision` argument validation for `mmwrite`

AtsushiSakai committed:
BUG: io: add `precision` argument validation for `mmwrite`

AtsushiSakai commented on scipy/io/_mmio.py:
```diff
@@ -20,6 +20,18 @@
+eps = np.finfo(np.float64).eps
+max_precision = np.floor(-np.log10(eps)).astype(int)
+"""
+MAX_PRECISION = 15
```
Should this parameters be the same across all environments? Or would it be better to calculate them dynamically?

h-vetinari replied:
I would make this 16. The standard eps for 64bit floats is around 2e-16, which is closer to 16 digits than 15.

ev-br committed:
Merge pull request #22170 from ev-br/spline_filters_array_api

ENH: signal: convert `symiirorder` and related filters to work with array backends

j-bowhay committed:
ENH: linalg.lapack: wrap ?sytri and ?hetri (#22228)

* ENH: linalg: wrap ?systri

* ENH: linalg: wrap ?hetri

* MAINT: update lda definition

dschmitz89 commented:
According to the issue, it works for up to 21 digits so far. Strictly speaking, raising the error for 15 already is a backwards compatibility break. How about only raising the error for the cases that otherwise cause the segmentation fault? One other thing is: should we validate that `precision` must be an integer?

AtsushiSakai commented:
Thank you for your comment. @dschmitz89 
The issue reports that the problem occurred when the precision = 22, but when I tested it using the latest code with unittests, I found that the integer overflow issue also occurred when the precision = 19.
```
scipy/io/tests/test_mmio.py:757: in test_precision
    mmwrite(self.fn, A, precision=precision)
        A          = <Dictionary Of Keys sparse array of dtype 'float64'
        with 1 stored elements and shape (10000000000000000001, 10000000000000000001)>
        n          = 10000000000000000001
        precision  = 19
        self       = <scipy.io.tests.test_mmio.TestMMIOCoordinate object at 0x11b77ed20>
        test_precisions = range(1, 20)
        test_values = [3.141592653589793, 1, 0.1, 0.01, 0.001, 0.0001, ...]
        value      = 3.141592653589793
scipy/io/_fast_matrix_market/__init__.py:528: in mmwrite
    a = a.tocoo()
        _fmm_core  = <module 'scipy.io._fast_matrix_market._fmm_core' from '/Users/atsushi_sakai/Library/CloudStorage/Dropbox/Program/python/scipy/build-install/lib/python3.12/site-packages/scipy/io/_fast_matrix_market/_fmm_core.cpython-312-darwin.so'>
        a          = <Dictionary Of Keys sparse array of dtype 'float64'
        with 1 stored elements and shape (10000000000000000001, 10000000000000000001)>
        comment    = None
        cursor     = <scipy.io._fast_matrix_market._fmm_core._write_cursor object at 0x11b7a89b0>
        field      = None
        precision  = 19
        symmetry   = 'general'
        target     = '/var/folders/s4/xj886tdn6wq0gytxmbgs98pw0000gn/T/tmpof9ficbp11230851/testfile.mtx'
scipy/sparse/_dok.py:446: in tocoo
    idx_dtype = self._get_index_dtype(maxval=max(self.shape))
        copy       = False
        nnz        = 1
        self       = <Dictionary Of Keys sparse array of dtype 'float64'
        with 1 stored elements and shape (10000000000000000001, 10000000000000000001)>
scipy/sparse/_base.py:1388: in _get_index_dtype
    return get_index_dtype(arrays,
        arrays     = ()
        check_contents = False
        get_index_dtype = <function get_index_dtype at 0x11a1e4ae0>
        maxval     = 10000000000000000001
        self       = <Dictionary Of Keys sparse array of dtype 'float64'
        with 1 stored elements and shape (10000000000000000001, 10000000000000000001)>
scipy/sparse/_sputils.py:312: in get_index_dtype
    maxval = np.int64(maxval)
E   OverflowError: Python int too large to convert to C long
        arrays     = ()
        check_contents = False
        int32max   = np.int32(2147483647)
        int32min   = np.int32(-2147483648)
        maxval     = 10000000000000000001
```

But, theoretically, a similar issue might happen when precision is between 16 and 18. This is the reason behind my suggestion to set the maximum precision to 15.
What do you think?

j-bowhay committed:
ENH: linalg: wrap ?langb (#22217)

* ENH: linalg: wrap ?langb

* Update scipy/linalg/flapack_gen_banded.pyf.src

Co-authored-by: Ilhan Polat <ilhanpolat@gmail.com>

---------

Co-authored-by: Ilhan Polat <ilhanpolat@gmail.com>

AustinZhang1024 committed:
MAINT: `sparse.linalg._isolve`: Remove postprocess function (#22064)

nickodell commented:
@j-bowhay Apologies, I don't have time to review this this week.

dschmitz89 commented:
> Thank you for your comment. @dschmitz89 The issue reports that the problem occurred when the precision = 22, but when I tested it using the latest code with unittests, I found that the integer overflow issue also occurred when the precision = 19.
> 
> ```
> scipy/io/tests/test_mmio.py:757: in test_precision
>     mmwrite(self.fn, A, precision=precision)
>         A          = <Dictionary Of Keys sparse array of dtype 'float64'
>         with 1 stored elements and shape (10000000000000000001, 10000000000000000001)>
>         n          = 10000000000000000001
>         precision  = 19
>         self       = <scipy.io.tests.test_mmio.TestMMIOCoordinate object at 0x11b77ed20>
>         test_precisions = range(1, 20)
>         test_values = [3.141592653589793, 1, 0.1, 0.01, 0.001, 0.0001, ...]
>         value      = 3.141592653589793
> scipy/io/_fast_matrix_market/__init__.py:528: in mmwrite
>     a = a.tocoo()
>         _fmm_core  = <module 'scipy.io._fast_matrix_market._fmm_core' from '/Users/atsushi_sakai/Library/CloudStorage/Dropbox/Program/python/scipy/build-install/lib/python3.12/site-packages/scipy/io/_fast_matrix_market/_fmm_core.cpython-312-darwin.so'>
>         a          = <Dictionary Of Keys sparse array of dtype 'float64'
>         with 1 stored elements and shape (10000000000000000001, 10000000000000000001)>
>         comment    = None
>         cursor     = <scipy.io._fast_matrix_market._fmm_core._write_cursor object at 0x11b7a89b0>
>         field      = None
>         precision  = 19
>         symmetry   = 'general'
>         target     = '/var/folders/s4/xj886tdn6wq0gytxmbgs98pw0000gn/T/tmpof9ficbp11230851/testfile.mtx'
> scipy/sparse/_dok.py:446: in tocoo
>     idx_dtype = self._get_index_dtype(maxval=max(self.shape))
>         copy       = False
>         nnz        = 1
>         self       = <Dictionary Of Keys sparse array of dtype 'float64'
>         with 1 stored elements and shape (10000000000000000001, 10000000000000000001)>
> scipy/sparse/_base.py:1388: in _get_index_dtype
>     return get_index_dtype(arrays,
>         arrays     = ()
>         check_contents = False
>         get_index_dtype = <function get_index_dtype at 0x11a1e4ae0>
>         maxval     = 10000000000000000001
>         self       = <Dictionary Of Keys sparse array of dtype 'float64'
>         with 1 stored elements and shape (10000000000000000001, 10000000000000000001)>
> scipy/sparse/_sputils.py:312: in get_index_dtype
>     maxval = np.int64(maxval)
> E   OverflowError: Python int too large to convert to C long
>         arrays     = ()
>         check_contents = False
>         int32max   = np.int32(2147483647)
>         int32min   = np.int32(-2147483648)
>         maxval     = 10000000000000000001
> ```
> 
> But, theoretically, a similar issue might happen when precision is between 16 and 18. This is the reason behind my suggestion to set the maximum precision to 15. What do you think?

Ok, probably have to test out myself locally to test the highest possible `precision`.

@nickodell : there is no time pressure to review this quickly. We are all volunteers, taking more than a week is no issue. :)

ev-br committed:
MAINT: interpolate: simplify complex c handling

ev-br committed:
MAINT: interpolate/BSpline: allocate the output in C

j-bowhay committed:
ENH: linalg.solve: use langb (#22251)

* ENH: linalg.solve: use langb

smammy committed:
ENH: io.wavfile: read unseekable files (#22208)

* ENH: read unseekable files in scipy.io.wavfile.

Allow scipy.io.wavfile.read() to read non-seekable files.

  * If passed stream is not seekable, wrap it in a SeekEmulatingReader,
    which tracks current position and implements tell() to report it.

Most seek()s in wavfile.read() simply seek forward from the current
position by a fixed amount. To address these:

  * Support limited SeekEmulatingReader.seek() that emulates seek()s
    that can be supported by read()ing and discarding the result.

Befor returning, read() attempts to seek back to the start of the file.
This is nice when possible, but if it isn't we shouldn't blow up. So:

  * Avoid final fid.seek(0) if underlying stream isn't seekable.

The remaining seek()s are in RF64 support, where chunk size is global
and stored in the RIFF header rather than in the chunk header.
_read_data_chunk() currently seeks back to a fixed offset from the file
origin to grab the chunk size. To address this:

  * Modify RF64 support to parse and store the chunk size in
    _read_riff_chunk() so that _read_data_chunk() doesn't have to seek()
    back for it. This applies whether or not reading from a seekable
    stream.

To test this, I added test_streams(), which ensures that wavfile reads
the same data no matter whether a path, seekable stream, or unseekable
stream is passed to read().

Fixes #11328

* TST: assert_array_equal -> assert_equal

---------

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

crusaderky committed:
TST: Mark with `xp` all tests in Array API-compatible modules

crusaderky committed:
TST: dev.py quietly ignores user markers

lucascolley committed:
Merge pull request #22261 from crusaderky/tst_markers

mdhaber committed:
Merge pull request #22262 from crusaderky/always_xp

TST: Mark with `xp` all tests in Array API-compatible modules

ev-br committed:
TST: signal: fix symiir tests (#22259)

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

crusaderky committed:
EHN: cluster: JAX support (non-jitted)

lucascolley committed:
Merge pull request #22255 from crusaderky/jax_cluster

crusaderky committed:
Merge pull request #22260 from crusaderky/skip_xp_backends

mdhaber committed:
MAINT: linalg.solve: raise when diagonal matrix is exactly singular (#22266)

* MAINT: linalg.solve: at least warn when diagonal matrix is exactly singular

* DOC: linalg.solve: more precise reason for raising LinAlgError

* MAINT: linalg: raise if rcond == 0

* MAINT: solve: pass 'lower' to sycon/hecon

---------

Co-authored-by: Robert Kern <robert.kern@gmail.com>

rgommers committed:
BLD: fix some issues with undeclared internal build dependencies

This is a follow-up to gh-21983, and fixes some more issues reported
on gh-18909. It's possible to run into these when building with
a high level of parallelism, or trying to build just a single
target like `spatial/_qhull.so` with Ninja.

crusaderky committed:
ENH: special: JAX support (non-jitted) (#22256)

rgommers committed:
TST: fix thread safety issue in interpolate.bsplines memmap test

This test was recently introduced in gh-22158. It causes a crash
when running the test suite with `pytest-run-parallel`, because of
writing to the same tempfile from multiple threads.

[skip cirrus] [skip circle]

ev-br committed:
Merge pull request #22272 from rgommers/fix-splines-test

TST: fix thread safety issue in interpolate.bsplines memmap test

j-bowhay committed:
ENH: linalg.solve: use gbcon

[skip ci]

crusaderky committed:
TST: clean up obsolete Array API fixtures (#22268)

* TST: clean up obsolete Array API fixtures

crusaderky committed:
_contains_nan for lazy arrays

lucascolley committed:
Merge pull request #22270 from rgommers/fix-more-ninjadeps

j-bowhay committed:
fix inequality and 1/0 based indexing dance

star1327p committed:
Replaced Twitter with X

star1327p committed:
Change the Twitter icon to X

Code source from https://github.com/pydata/pydata-sphinx-theme/blob/main/docs/conf.py

star1327p committed:
Removed Twitter or X entirely in the Code of Conduct

mdhaber committed:
DOC: optimize.curve_fit: add note about more advanced curve fitting (#22269)

* DOC: optimize.curve_fit: add note about more advanced curve fitting

[docs only]

mdhaber committed:
TST: stats.Normal: bump tolerance on test of logcdf (#22276)

* TST: stats.Normal: bump tolerance on test of logcdf

* Update scipy/stats/tests/test_continuous.py

[skip ci]

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

---------

Co-authored-by: Daniel Schmitz <40656107+dschmitz89@users.noreply.github.com>

crusaderky committed:
Merge pull request #22280 from crusaderky/tst_skip

crusaderky committed:
TST: special: better skip message for stdtrit on JAX (#22281)

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

tupui committed:
Merge pull request #22277 from star1327p/twitter-to-x

ev-br committed:
TST: Add skips for GPU CI failures (#22279)

rgommers committed:
CI: add a GPU CI job (#22275)

dschult committed:
MAINT: sparse: better error message for unsupported dtypes (#22284)

mdhaber committed:
Merge pull request #22274 from crusaderky/nan_policy

ENH: `_contains_nan` for lazy arrays

czgdp1807 committed:
BUG: Fix `Akima1DInterpolator` by returning linear interpolant for `y.shape[0] == 2` (#22278)

* BUG: interpolate: Akima1DInterpolator: use linear interpolation if there are only 2 points

reviewed at https://github.com/scipy/scipy/pull/22278

rgommers committed:
CI: fix skip/trigger condition of GPU CI job (#22289)

[skip circle] [skip cirrus]

MatthewFlamm committed:
BUG: Fix banded Jacobian for lsoda: `ode` and `solve_ivp` (#22283)

* Add failing test for banded Jacobian with lsoda

* lint: trailing newline

* remove print

* use wrapper for banded jacobian for lsoda.

* Use `assert_allclose`

* remove vode comment

* add comment about lsoda for original test

* Make comment style consistent

---------

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

mdhaber committed:
Merge pull request #22273 from j-bowhay/solver_banded_con

ENH: linalg.solve: use gbcon

HaoZeke committed:
MAINT: Update highs subproject commit

rgommers committed:
Merge pull request #22292 from HaoZeke/updHighsCommit

MAINT: Update highs subproject commit

Gautzilla committed:
DOC: signal.ShortTimeFFT.nearest_k_p: fix typo (#22295)

SL3333PY committed:
MAINT: remove end year from copyright

tupui committed:
Merge pull request #22300 from SL3333PY/copyright

Kai-Striega committed:
BENCH: sparse.csgraph.dijkstra: add benchmark (#22237)

* Add benchmark discussed in gh-20717

gh-20717 was the first of multiple planned improvements to SciPy. We should
track perf changes appropriately for this.

* STY: Fix line length

We need the line length to be <88 chars

[lint only]

* BENCH/ENH: sparse.csgraph.dijkstra: Use new random

This commit uses the ``sparse.random_array`` interface rather than the
 (now outdated) ``sparse.rand`` interface.

 [skip actions] [skip cirrus]

* BENCH: sparse.csgraph.dijkstra: Skip n=1000

* BENCH: sparse.csgraph.dijkstra: use data_sampler

[skip actions] [skip cirrus]

* Removed unused import

[skip ci]

---------

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

SL3333PY committed:
MAINT: remove unused library import missed in PR #22300

tupui committed:
Merge pull request #22302 from SL3333PY/copyright

tylerjereddy committed:
MAINT, DOC: forward port 1.15.1 relnotes

* Forward port the SciPy `1.15.1` release notes and
update the version switcher.

[docs only]

OmerCohen7530 committed:
DOC: Clarify support behavior in rv_continuous documentation (#22201)

* DOC: Clarify support behavior in rv_continuous documentation

* Update _distn_infrastructure.py

Fix doctest output for rv_continuous example to match expected output format

- Adjusted expected output to match actual behavior of uniform distribution with loc and scale parameters.
- Updated the doctest to reflect the correct bounds representation without `np.float64`.

* MAINT: stats: reset to main

* DOC: stats: document a/b attributes of rv_continuous/rv_discrete and show example use of 'support'

* Apply suggestions from code review

[docs only]

* Update scipy/stats/_distn_infrastructure.py

[docs only]

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

* DOC: stats: fix remaining issues per review

[docs only]

---------

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>
Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

akhmerov committed:
ENH: stats.special_ortho_group: speed up, allow 1x1 and 0x0 ortho and unitary groups (#22304)

* ENH: streamline implementation of stats.special_ortho_group
* ENH: allow 1x1 and 0x0 ortho and unitary groups
* TST: switch from random.seed to random.default_rng

andfoy committed:
BUG: immortalize uarray global strings in order to prevent negative refcounts (#22038)

#### Reference issue
Fixes https://github.com/scipy/scipy/issues/21214
Closes https://github.com/scipy/scipy/pull/21218

#### What does this implement/fix?
This PR addresses the segfaults caused by the occurrence of negative refcounts when uarray static strings were being released when the interpreter didn't exist.

---------

Co-authored-by: peterbell10 <peterbell10@live.co.uk>

lithomas1 committed:
TST: replace `pytest.xfail` with `skip/xfail_xp_backends` (#22311)

mdhaber committed:
MAINT: stats.Mixture: make return type consistent

tupui committed:
Merge pull request #22322 from mdhaber/MixtureScalar

crusaderky committed:
TST: array types: wrap namespaces centrally (#22320)

dschmitz89 commented:
On my Mac, I can write with `precision=200` without any issues. This leaves me very confused: the user report stated that issues started at 22, for @AtsushiSakai  they started at 19 while for me they do not occur at all. Is this hardware dependent as a consequence of the C++ source?

ev-br committed:
ENH: signal.upfirdn: array API standard support (#22226)

AtsushiSakai committed:
ENH: io: add invalid field name warning for `savemat` (#22324)

* ENH: io: add invalid field name warning for `savemat`

* ENH: io: fix doc string typo

* ENH: io: fix lint CI error

* ENH: io: fix lint CI error

* STY: style nits

---------

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

davidkun committed:
ENH: spatial.transform.Rotation: Add __repr__ method to (#22293)

* ENH: Add __repr__ method to scipy.spatial.transform.Rotation

* ENH: Rotation __repr__ uses .from_quat() (PR feedback)

* ENH: Rotation __repr__ uses .from_matrix() (PR feedback)

ev-br committed:
MAINT: interpolate: simplify complex.view(float)

j-bowhay committed:
Merge pull request #22264 from ev-br/bspline_out_mutable

MAINT: interpolate: make BSpline allocate out arrays in C

crusaderky committed:
TST: `_lib`: run tests with `@jax.jit` (#22308)

j-bowhay committed:
MAINT: linalg: reorganize tridiagonal eigenvalue routines

ev-br committed:
MAINT: signal.oaconvolve: avoid xp <-> numpy conversions (#22122)

* MAINT: signal: rm np.split, use array API primitives instead

* MAINT: signal: draft xp_pad for oaconvolve

* TST: signa: test oaconvolve on CUDA

* MAINT: fft: remove a stray np.asarray call

* DOC: fft: move the docstring from fft/_helper.py to fft/_pocketff/helper.py

The docstring is the same.

* MAINT: signal: xp_pad -> xpx.pad

* MAINT: signal: linter

* MAINT: add a code comment

* update comment [skip ci]

---------

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

cbouss committed:
do not check dtype in test_compare_with_GCVSPL

mdhaber committed:
Merge pull request #22340 from j-bowhay/trieig

MAINT: linalg: reorganize tridiagonal eigenvalue routines

ev-br committed:
Merge pull request #22345 from cbouss/test_compare_with_GCVSPL_endianness

TST: turn off dtype check due to endianness

WarrenWeckesser committed:
BUG: special: Fix unchecked memory allocations in `specfun.h` (#22080)

* WIP: Fix unchecked memory allocation in aswfa()

Summary of the changes:

* Add another SF error code, "SF_ERROR_NOMEM", that is used by ufuncs
  that require internally allocated memory.  Just like the other possible
  errors, how this error is handled can be queried and set using `geterr()`
  and `seterr()`.  Unlike the other error conditions, the default behavior
  is to raise an exception if a memory allocation fails.

* Modify the function `aswfa()` in `specfun.h` to return a status code
  that indicates when a mempory allocation failed.

* Modify the functions `prolate_aswfa_nocv()`, `oblate_aswfa_nocv()`,
  `prolate_aswfa()` and `oblate_aswfa()`  in `sphd_wave.h` to check
  the return value of `aswfa()`, and set the error state and return values
  appropriately if `aswfa()` returns `NoMemory`.

* MAINT: Rename 'nomem' -> 'memory'.

* Fix all uses of unchecked allocations in specfun.h

* Fix aswfa() in specfunc.h

* Fix refguide check.

* Add comment about the return value when nonconvergence is detected in rmn2l.

* Change (c/m)alloc calls to use 'new' wrapped in a unique_ptr.

* Fix a few more names passed to set_error()

* Set outputs to nan in mtu12 when alloc fails
Also update docstring-like comments of some C++ functions.

anntzer committed:
DOC: Harmonize summary line of docstrings of iterative sparse solvers. (#22348)

* DOC: Harmonize summary line of docstrings of iterative sparse solvers.

The summary lines are shown next to one another under "Solving linear
problems" in the docs for sparse.linalg; give them all the same format
and also clarify which methods require symmetry or
positive-definiteness.

---------

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

larsoner committed:
BUG: Fix bug with dpss degenerate case (#22344)

* BUG: Fix bug with dpss degenerate case

* FIX: Arg

ParamSingh24 committed:
DOC: sparse.linalg.eigsh: fix inconsistent definitions of OP and OPinv#18214 (#22351)

* solved the issue: inconsistent definitions of OP and OPinv in eigsh #18214

* Apply suggestions from code review

[docs only]

---------

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

ev-br committed:
TST: signal: update to dpss updates

j-bowhay committed:
TST: interpolate: small tolerance bump to TestAAA.test_basic_functions

dschmitz89 commented:
For me the error also only occurs for `precision>=22`. Is this maybe hardware dependent? Raising the error for 16 and higher is likely the right approach though, more precision will anyway not be written. Would need a deprecation cycle in my opinion though. Thoughts @AtsushiSakai  ?

ev-br committed:
Merge pull request #22356 from j-bowhay/gh-22355

TST: interpolate: small tolerance bump to TestAAA.test_basic_functions

ev-br committed:
MAINT: signal/windows: f64 degenerate windows, for consistency

HaoZeke committed:
MAINT: Stop installing libhighs

Co-authored-by: drew-parsons <drew-parsons@users.noreply.github.com>

Closes #22349

lucascolley committed:
Merge pull request #22357 from HaoZeke/noHighsLib

jorenham committed:
MAINT: remove temporary `# type: ignore`'s from #22162

lucascolley committed:
Merge pull request #22359 from jorenham/maint/22164

DietBru committed:
Small code formatting improvement in scipy/signal/_fir_filter_design.py

[skip ci]

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

DietBru committed:
Merge pull request #18375 from sagi-ezri/add-fwind1-filter

ENH: signal: Add `firwin_2d` filter

roryyorke committed:
BUG: signal.bilinear handles complex input, and strips leading zeros (#22313)

* Add test for signal.bilinear for complex input

* BUG: signal.bilinear handles complex input

Fixes gh-13823.

* Add leading-zeros test for signal.bilinear

* BUG: trim leading zeros in inputs in signal.bilinear

* Rework of `signal.blinear` function.

* Simplified algorithm by utilizing NumPy's Polynomial class.
* Improved docstr and example.
* Added unit test to check for exceptions.

* Improve docstr of `signal.blinear` function by making the formulas a narrower by using subscript array notation. Also, some typos were removed.

* Reduce the width of the longest equation in the  docstr of `signal.blinear` a little more.

* Remove typos in function signal.bilinear

Co-authored-by: Rory Yorke <rory.yorke@gmail.com>

* Correct two typos in docst of `signal.bilinear`.

---------

Co-authored-by: Dietrich Brunn <Dietrich.Brunn@web.de>

AtsushiSakai committed:
BUG: io: using deprecation error instead of throwing exception

AtsushiSakai committed:
BUG: io: using deprecation error instead of throwing exception

j-bowhay commented on scipy/io/_mmio.py:
```diff
@@ -20,6 +21,20 @@
+
+def _validate_precision(precision):
+    if (precision is not None) and not (1 <= precision <= MAX_PRECISION):
+        msg = (f"From SciPy 1.18, an exception will be thrown if the "
```
```suggestion
        msg = (f"From SciPy 1.18.0, an exception will be thrown if the "
```

j-bowhay commented on scipy/io/tests/test_mmio.py:
```diff
@@ -761,6 +761,13 @@ def test_precision(self):
+    def test_invalid_precision(self, precision):
+        A = scipy.sparse.dok_array((2, 2))
+        with pytest.warns(FutureWarning,
+                          match='From SciPy 1.18, an exception will be thrown'):
```
```suggestion
                          match='From SciPy 1.18.0, an exception will be thrown'):
```

j-bowhay commented on scipy/io/_mmio.py:
```diff
@@ -149,6 +164,10 @@ def mmwrite(target, a, comment='', field=None, precision=None, symmetry=None):
         Either 'real', 'complex', 'pattern', or 'integer'.
     precision : None or int, optional
         Number of digits to display for real or complex values.
+    .. warning::
```
```suggestion
    .. deprecated:: 1.16.0
```

j-bowhay commented on scipy/io/_mmio.py:
```diff
@@ -149,6 +164,10 @@ def mmwrite(target, a, comment='', field=None, precision=None, symmetry=None):
     precision : None or int, optional
         Number of digits to display for real or complex values.
+    .. warning::
+        From SciPy 1.18, an exception will be thrown if the precision
```
```suggestion
        From SciPy 1.18.0, an exception will be thrown if the precision
```

j-bowhay commented on scipy/io/_fast_matrix_market/__init__.py:
```diff
@@ -389,6 +389,10 @@ def mmwrite(target, a, comment=None, field=None, precision=None, symmetry="AUTO"
         Either 'real', 'complex', 'pattern', or 'integer'.
     precision : None or int, optional
         Number of digits to display for real or complex values.
+    .. warning::
```
```suggestion
    .. deprecated:: 1.16.0
```

j-bowhay commented on scipy/io/_fast_matrix_market/__init__.py:
```diff
@@ -389,6 +389,10 @@ def mmwrite(target, a, comment=None, field=None, precision=None, symmetry="AUTO"
     precision : None or int, optional
         Number of digits to display for real or complex values.
+    .. warning::
+        From SciPy 1.18, an exception will be thrown if the precision
```
```suggestion
        From SciPy 1.18.0, an exception will be thrown if the precision
```

j-bowhay committed:
TST: bump tolerance on TestHyp2f1.test_region3[hyp2f1_test_case23] (#22364)

AtsushiSakai committed:
Update scipy/io/_mmio.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/_mmio.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/tests/test_mmio.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/tests/test_mmio.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/_mmio.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/_mmio.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/_mmio.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/_mmio.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/_fast_matrix_market/__init__.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/_fast_matrix_market/__init__.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/_fast_matrix_market/__init__.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/_fast_matrix_market/__init__.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

dschmitz89 commented:
Thanks @AtsushiSakai , this looks close. I still think that while we are at it we should validate that `precision` is an int.

j-bowhay committed:
Merge pull request #21783 from ev-br/pr/20668

ENH: signal.windows: add array API support (take 2)

j-bowhay commented on scipy/io/_mmio.py:
```diff
@@ -20,6 +21,20 @@
+    if (precision is not None) and not (1 <= precision <= MAX_PRECISION):
+        msg = (f"From SciPy 1.18.0, an exception will be thrown if the "
+               f"precision input is outside the range of 1 to {MAX_PRECISION-1}")
+        warnings.warn(msg, FutureWarning, stacklevel=2)
```
```suggestion
        warnings.warn(msg, FutureWarning, stacklevel=3)
```

AtsushiSakai committed:
Update scipy/io/_mmio.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/_mmio.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
BUG: io: Add integer check for precision argument

AtsushiSakai committed:
BUG: io: Add integer check for precision argument

dschult committed:
BUG: sparse.linalg.norm: fix return type (#22372)

cupy reports https://github.com/cupy/cupy/issues/8866 that
sparse.linalg.norm returns a python number instead of a 0-dim numpy array

dschult committed:
DOC: sparse.linalg: sparray updates in doc_strings and Sakurai and MikotaK methods (#22376)

j-bowhay committed:
DOC: interpolate.AAA: add may vary to example 

[docs only]

lucascolley committed:
Merge pull request #22379 from j-bowhay/aaadoctest

j-bowhay committed:
BUG: special: Fix unchecked malloc in stirling2.h (#22339)

Use std::unique_ptr and new (std::nothrow) instead of malloc.

Closes gh-22336.

yahya-darman committed:
DOC: integrate: fix quad documentation to correctly describe QAGS algorithm (#22366)

* DOC: Fix quad documentation to correctly describe QAGS algorithm

The documentation incorrectly stated that quad uses Clenshaw-Curtis method
for finite integration limits. Updated to correctly reflect that QAGS uses
21-point Gauss-Kronrod quadrature with epsilon algorithm acceleration.

---------

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

DietBru committed:
DOC: Replace link to X in header with link to scientific python forum (#22380)

* DOC: Replace link to X in header with link to scientific python forum in header.

* Moved file `custom-icons.js` from `doc/source/` into `doc/source/_static` to make it useful.

dschult committed:
DOC: sparse.linalg: fix doctest in scipy.sparse.linalg._norm.py (#22375)

* fix doctest in scipy.sparse.linalg._norm.py

* change tol on doctest and remove maxiter kwd [docs only]

WarrenWeckesser committed:
MAINT: special: A bit of clean up in stirling2.h (#22381)

* MAINT: special: A bit of clean up in stirling2.h

* We've been compiling this as C++ for a while, and we now use
  std::unique_ptr and new, so there is no need for the use of
  `#if defined(__cplusplus)`.
* Remove some unused includes.
* Use a C++-style cast.
* Make a few minor style and whitespace corrections.

* MAINT: special: Qualify all cmath functions in stirling2.h with 'std::'.

Transurgeon committed:
STY: fix lint rule UP031 (#21987)

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

j-bowhay commented on scipy/io/_mmio.py:
```diff
@@ -31,10 +31,14 @@
-        warnings.warn(msg, FutureWarning, stacklevel=3)
+    if precision is not None:
+        if not isinstance(precision, int):
+            raise TypeError("Precision value must be an integer")
```
What about `np.int32` etc did this previously work and does it still work?

AtsushiSakai replied:
Oh, great catch. I fixed and added tests.

j-bowhay replied:
I wonder if we should use `operator.index` it seems to used quite a bit elsewhere to validate ints eg
https://github.com/scipy/scipy/blob/261784e0892ce3ceeafddaa6ced58c37eec12283/scipy/fftpack/_helper.py#L46

AtsushiSakai replied:
Thank you very much. I have modified it to use `operator.index`. With this change, `np.array` wrapped an integer can now be used as precision values. 

```
In [16]: operator.index(np.array(1))
Out[16]: 1
```

If other functions are performing the same check, I believe this is good for consistency.

j-bowhay reviewed (commented):
Thanks @AtsushiSakai could you post on the forum about the new future warning to check everyone is happy

czgdp1807 committed:
DOC: Fix documentation for ``predecessors`` matrix in ``shortest_path``, ``bellman_ford``, ``johnson`` and ``dijkstra`` (#22388)

---------

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
BUG: io: add numpy integer handling.

AtsushiSakai committed:
BUG: io: add numpy integer handling.

AtsushiSakai commented:
Sorry, let me confirm. "the fourm" means discord thread?

j-bowhay commented:
> Sorry, let me confirm. "the fourm" means discord thread?

https://discuss.scientific-python.org/c/contributor/scipy the replacement to the old mailing list:)

AtsushiSakai commented:
Thansk. Done: https://discuss.scientific-python.org/t/dep-deprecate-16-precision-argument-for-scipy-io-mmwrite/1688

evbernardes committed:
DOC: spatial: Added mention of Davenport Angles to Rotation class documentation (#22397)

Added line mentioning "Davenport Angles (Generalized Euler Angles)"

dschult committed:
DOC: sparse.linalg: add two recent functions to namespace and fix doctests (#22374)

* fix doctests and make two recent functions visible

* update refguide

[docs only]

* fix capitalization   [docs only]

Co-authored-by: Daniel Schmitz <40656107+dschmitz89@users.noreply.github.com>

---------

Co-authored-by: Daniel Schmitz <40656107+dschmitz89@users.noreply.github.com>

dschult committed:
revert NotImplemented return values in dot & others (#22373)

AtsushiSakai committed:
DOC: Add "Assert function selection guideline" doc in the new page "Tips for writing tests". (#22389)

AtsushiSakai committed:
BUG: io: using operator.index

AtsushiSakai committed:
BUG: io: using operator.index

AtsushiSakai committed:
BUG: io: using operator.index

AtsushiSakai committed:
BUG: io: using operator.index

rgommers committed:
CI: build Linux aarch64 wheels on GitHub Actions instead of Cirrus CI

This unifies the wheels builds, which is a lot easier on maintenance.

In the process, did the following:
- update to setup-miniconda 3.1.1. for GHA Linux arm64 runner support
- add `conda-remove-defaults` to `setup-miniconda`

rgommers committed:
TST: remove thread-unsafe skips for a now fixed Cython fused types issue

This was Cython issue 6506, fixed a while back. The Cython nightly wheels
and the latest conda-forge package (3.1.0a2.dev0) contain the fix.

[skip ci]

dschmitz89 reviewed (approved):
Thanks @AtsushiSakai .

DietBru committed:
ENH: signal.ShortTimeFFT: determine arbitrary dual windows (#20610)

* ENH: Determine arbitrary dual windows for short-time Fourier Transform

Provide a new function `closest_STFT_dual_window()` to calculate the closest dual window in a least-squares sense for a given desired window as well as a new method `ShortTimeFFT.from_win_equals_dual()` to determine windows which are equal to their dual window.

* Miscellaneous documentation improvements

* Fixed linter warnings

* Replace `assert_allclose` with `xp_assert_close` in `test_short_time_fft.py`

* Added unitary scaling to ShortTimeFFT and various documentation improvements mainly about unitary mappings.

Also:
* Some fixed double backticks in docstrs
* Insured that L2-norm typesetting corresponds to Wikipedia's typesetting convention.
* Further small documentation improvements

* In file `_short_time_fft.py`:
* Make the default of parameter `desired_dual` in function `ShortTimeFFT.closest_STFT_dual_window` a rectangular window.
* Improve docstr of `ShortTimeFFT.from_win_equals_dual`.
* Add "Parameters" and "Returns" section to the docstrs of `ShortTimeFFT.stft_detrend` and `ShortTimeFFT.spectrogram`.
* Add types to "Parameters" section in `ShortTimeFFT.stft`, `ShortTimeFFT.stft_detrend` and `ShortTimeFFT.spectrogram`.
* Make the example in the `ShortTimeFFT.spectrogram` docstr more copy and paste friendly.

In file `test_short_time_fft`:
* Added test for testing detrending in `ShortTimeFFT.stft_detrend`.
* Add clarifying comment to `test_closest_STFT_dual_window_cola` test.

Furthermore:
* Improve legacy message in the docstr of `signal.check_COLA`.
* Shorten width of Eq. `eq_STFT_CanonDualWin` in file `signal.rst`.

* Increased tolerance in `test_compare_stft_detrend` due to CI failing.

---------

Co-authored-by: Eric Larson <larson.eric.d@gmail.com>

rgommers committed:
MAINT: special: clean up os/warnings modules exposed in special namespace (#22398)

tylerjereddy committed:
Merge pull request #22399 from rgommers/remove-skip-fusedtypes

TST: remove thread-unsafe skips for a now fixed Cython fused types issue

tylerjereddy committed:
Merge pull request #22305 from tylerjereddy/treddy_forward_1.15.1_notes

MAINT, DOC: forward port 1.15.1 relnotes

jorenham committed:
TYP: Generic ``sparray`` and ``spmatrix`` types in ``scipy.sparse``

lucascolley committed:
Merge pull request #22401 from jorenham/typing/generic-sparse-arrays

mdhaber committed:
MAINT: integrate.cumulative_simpson: bump test tolerance

lucascolley committed:
Merge pull request #22408 from mdhaber/cumulative_simpson_tolerance

mdhaber committed:
DOC: stats.gaussian_kde: clarify Silverman method (#22411)

[docs only]

mdhaber committed:
MAINT: stats.zmap: restore support for complex data (#22405)

* MAINT: stats.zmap: restore support for complex data

* MAINT: stats.zmap: fix masked array support

* Apply suggestions from code review

PTNobel committed:
BUG: sparse: fix selecting wrong dtype for coo coords (#22353)

* Fixes bug with selecting wrong dtype for coo coords

* add test for reshape having wrong dtype

* test that smaller dtype is maintained even if intermediate values are big

---------

Co-authored-by: Dan Schult <dschult@colgate.edu>

yahya-darman committed:
DOC: signal: Use where='post' when plotting discrete response #21020 (#22396)

* Update documentation for dstep and dimpulse to use where='post'

* Update _ltisys.py; took out label='(where="post")'

lucascolley committed:
Merge pull request #22407 from lucascolley/issparse

mdhaber committed:
MAINT: stats.Mixture: fix inverse functions when mean is undefined (#22337)

* MAINT: stats.Mixture: fix inverse when mean is undefined

* Apply suggestions from code review

* Update scipy/stats/_distribution_infrastructure.py

star1327p committed:
DOC: stats: Edited the NIST Handbook reference (#22413)

* DOC: Edited the NIST Handbook reference

---------

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

yahya-darman committed:
DEP: optimize.nnls: deprecate unused atol parameter (#22386)


---------

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

j-bowhay commented on scipy/io/_mmio.py:
```diff
@@ -20,6 +22,24 @@
+        if not (1 <= precision <= MAX_PRECISION):
+            msg = (f"From SciPy 1.18.0, an exception will be thrown if the "
+                   f"precision input is outside the range of 1 "
+                   f"to {MAX_PRECISION-1}")
```
```suggestion
                   f"to {MAX_PRECISION}")
```
The check is not a strict inequality?

AtsushiSakai replied:
Good catch!!

rgommers committed:
TST: linalg: bump tolerances in two TestBatch tests

Closes gh-22415

[skip ci]

mdhaber committed:
Merge pull request #22416 from rgommers/fix-linalg-batch-failures

TST: linalg: bump tolerances in two TestBatch tests

tylerjereddy committed:
BUG: cluster: `cophenet` intercept invalid linkage matrix count (#22187)

* BUG: catch invalid linkage count

* Fixes gh-22183.

* The input data from gh-22183 causes an out
of bounds memory access in a 1-D `memoryview`
in the Cython `cophenetic_distances` function.
Prevent this by enforcing a check for an allowable
upper bound on the cluster membership (4th column)
of the linkage matrix `Z` received by `is_valid_linkage`.

Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>

* TST: PR 22187 revisions

* Adjust `test_gh_22183()` such that the linkage matrix is invalid
only because of an excessive observation count, rather than
also being invalid because the large observation count is
not integral.

* Adjust `test_gh_22183()` such that is properly flushes
through array API backends.

* DOC: backticks

---------

Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>
Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/_mmio.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

AtsushiSakai committed:
Update scipy/io/_mmio.py

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

crusaderky committed:
TST: use singular `reason=` in `skip_xp_backends`

lucascolley committed:
Merge pull request #22420 from crusaderky/reason

crusaderky committed:
MAINT: `_lib`: adapt `array_namespace` to accept scalars with arrays (#22422)

WarrenWeckesser committed:
BUG: special: Fix a memory leak in the AMOS function besy(). (#22423)

Use a unique_ptr to manage the heap-allocated array.

Closes gh-22314.

rgommers committed:
DOC: add note on getting the version switcher to behave to release docs

Follows up on the issue discussed in gh-22305

[docs only]

crusaderky committed:
BUG: ndimage: `binary_erosion` vs. broadcasted input (#22421)

crusaderky committed:
MAINT: cluster: vectorize tests in `is_valid_linkage`

lucascolley committed:
ENH: array types: add dask.array support (#22240)

Co-authored-by: Guido Imperiale <crusaderky@gmail.com>
Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>
Co-authored-by: Thomas Li <47963215+lithomas1@users.noreply.github.com>

sarthakg004 committed:
DOC: Corrected formatting in integrate tutorial (#22431)

* Corrected formatting

* Update doc/source/tutorial/integrate.rst

[docs only]

---------

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

tylerjereddy committed:
Merge pull request #22428 from rgommers/doc-versionswitcher

DOC: add note on getting the version switcher to behave to release docs

rgommers committed:
Merge pull request #22430 from crusaderky/is_valid_linkage

MAINT: cluster: vectorize tests in `is_valid_linkage`

czgdp1807 committed:
DOC: Add example to show usage of ``predecessors`` matrix returned from `sparse.csgraph.shortest_path` (#22387)

* DOC: Added example to show usage of predecessors matrix for constructing shortest paths

* MISC: Remove trailing space

* DOC: Add comments in the loop constructing shortest paths

* DOC: Linked Word Ladders example in shortest_path documentation

* DOC: Removed redundant example from shortest_path documentation

* DOC: Add commentary to explain shortest path reconstruction

* [docs only]

crusaderky committed:
MAINT: bump to array-api-extra git tip (#22435)

j-bowhay committed:
DOC: integrate.tanhsinh: remove incorrect reference to _differentiate (#22434)

steppi committed:
MAINT: special: Update handling of `betainc` and `betaincc` edge cases (#22425)

bewygs committed:
DOC: optimize.dual_annealing: Fix year of publication (#22440)

steppi committed:
MAINT: special: Add `log1mexp` for ``log(1 - exp(x))`` (#22439)

* MAINT: special: add private function _log1mexp
* TST: special: add tests for log1mexp
* MAINT: stats: replace Python _log1mexp with ufunc

---------

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

sarthakg004 committed:
DOC: Modified Link for code of conduct documentation (#22442)

sarthakg004 committed:
DOC: corrected path to code of conduct (#22443)

steppi committed:
BUG: special: Fix incorrect handling of `nan` input in `gammainc` and `gammaincc` (#22441)

* BUG: Add nan checks in gammainc and gammaincc scalar kernels

* TST: Add tests for nan inputs for gammainc(c)

* BUG: fix chdtrc out of domain case

* TST: Add nan arg tests for chdtr(c)

rgommers committed:
CI: avoid mpmath pre-release version that's failing in CI (#22445)

xref gh-22395

rgommers committed:
CI: migrate Linux aarch64 job to GitHub Actions

ev-br committed:
MAINT: stats.multivariate_normal.cdf: replace Fortran `mvnun` with Cython (#22298)

MAINT: stats.multivariate_normal.cdf: replace Fortran `mvnun` with Cython

---------

Co-authored-by: Robert Kern <rkern@enthought.com>

rgommers committed:
CI: remove Cirrus CI configuration [wheel build]

lucascolley committed:
Merge pull request #22446 from rgommers/gha-aarch64

matthew-brett committed:
BUG: io.loadmat: throw error on file containing all zeros (#22447)

scottshambaugh committed:
BUG: scipy.spatial: Fix inaccurate orthonormalization in `Rotation.from_matrix()` (#22418)

* BUG: fix inaccurate orthonormalization in Rotation.from_matrix()

* Code review updates

---------

Co-authored-by: Scott Shambaugh <scottshambaugh@users.noreply.github.com>

mdhaber committed:
Merge pull request #21684 from j-towns/relax-dm-constraint

MAINT: `stats.dirichlet_multinomial` relax `n` to `>= 0`

j-bowhay committed:
ENH: linalg: wrap ?stevd (#22426)

mdhaber committed:
DOC: optimize.elementwise.find_minimum: fix documented termination condition (#22448)

[docs only]

ev-br committed:
ENH: signal.vectorstrength: add array API standard support (#22008)

* ENH: signal: TestVectorstrength array api + its tests

* MAINT: signal: simplify vectorstrength array api

* TST: Apply suggestions from code review

DietBru committed:
DOC: Improve docstrs of `dlsim`, `dimpulse`, `dstep`, `dfreqresp` and `dbode` in file `_ltisys.py`. (#22453)

* DOC: Improve docstrs of `dlsim`, `dimpulse`, `dstep`, `dfreqresp` and `dbode` in file `_ltisys.py`.

* Made the description of the parameter `system` more consistent.
* Improved plots plots a little bit.

* DOC: Fixed test position in plot of `signal.dfrqresp`.

mdhaber committed:
ENH: stats.gstd: add array API support (#22455)

* ENH: stats.gstd: add array api support

* MAINT: stats.gstd: add integer support for array_api_strict

* ENH: stats.gstd: add keepdims, nan_policy support

* STY: stats.gstd: linting

* TST: stats.gstd: skip dask.array because lazywhere doesn't work yet

DietBru committed:
BUG: signal.ShortTimeFFT: make attributes `win` and `dual_win` immutable (#22454)

* BUG: signal.ShortTimeFFT: make attributes win and dual_win immutable

Also, some minor docstr issues were fixed.

dschmitz89 commented:
@j-bowhay : are you happy with this?

tylerjereddy commented on scipy/io/_fast_matrix_market/__init__.py:
```diff
@@ -389,6 +389,10 @@ def mmwrite(target, a, comment=None, field=None, precision=None, symmetry="AUTO"
         Number of digits to display for real or complex values.
+    .. deprecated:: 1.16.0
+        From SciPy 1.18.0, an exception will be thrown if the precision
+        input is outside the range of 1 to 15, because these valuas are
```
minor comment as I'm catching up on GH notifications: `valuas` -> `values` I think

tylerjereddy commented on scipy/io/_mmio.py:
```diff
@@ -149,6 +169,10 @@ def mmwrite(target, a, comment='', field=None, precision=None, symmetry=None):
         Number of digits to display for real or complex values.
+    .. deprecated:: 1.16.0
+        From SciPy 1.18.0, an exception will be thrown if the precision
+        input is outside the range of 1 to 15, because these valuas are
```
same here

j-bowhay commented:
> @j-bowhay : are you happy with this?

No objections but I haven't looked into the issue enough to "approve" either

AtsushiSakai committed:
DEP: io: fix typo

AtsushiSakai committed:
DEP: io: fix typo

AtsushiSakai committed:
Merge remote-tracking branch 'origin/issue19949' into issue19949

h-vetinari commented on scipy/io/_mmio.py:
```diff
@@ -20,6 +22,24 @@
+
+def _validate_precision(precision):
+    if precision is not None:
+        # check the input can be converted to an integer
+        precision = operator.index(precision)
```
This is an unusual way (for me at least) to test convertability to an integer. Can we do something less magical?

j-bowhay replied:
This is quite a standard way of doing things across SciPy so unless there's a really nice way I'm not sure we should break the norm https://github.com/search?q=repo%3Ascipy%2Fscipy%20operator.index&type=code

h-vetinari reviewed (commented):
Basically LGTM, but some small nits

AtsushiSakai committed:
DEP: io: Increase maximum precision support from 15 to 16.

Updated `MAX_PRECISION` and related validations in MMIO and tests to support a precision range of 1 to 16. Adjusted corresponding docstrings and parameterizations to reflect the new upper limit.

AtsushiSakai committed:
Merge remote-tracking branch 'origin/issue19949' into issue19949

# Conflicts:
#	scipy/io/_fast_matrix_market/__init__.py
#	scipy/io/_mmio.py
#	scipy/io/tests/test_mmio.py

rgommers commented:
@AtsushiSakai it looks like you had a rebase hiccup here

AtsushiSakai commented:
Sorry. Let me close this PR once, then recrete my environment and this PR again.

ev-br commented:
Did you have a chance to fix the rebase @AtsushiSakai ? The linked issue is tagged for 1.16, so it'd be great to either make it into 1.16 or --- if it misses the imminent branching --- just after branching, so that it has time to brew in the 1.17 cycle.

AtsushiSakai commented:
@ev-br sorry. My dev environment is broken, and i don’t have time to fix it. So, is it possible to change the milestone to 1.17? I will retry it.

ev-br commented:
Of course, thanks for clarifying! 
I'll bump the milestone on the issue.
